torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1438,6 +1438,116 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
   return work;
 }
 
+template <typename Fn, typename PreProcess, typename PostProcess>
+c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::oneAndAll(
+    std::vector<at::Tensor>& tensor,
+    std::vector<std::vector<at::Tensor>>& tensor_list,
+    Fn fn,
+    PreProcess pre,
+    PostProcess post,
+    OpType opType,
+    const char* profilingTitle) {
+
+  errorIfCapturingNonCapturableNCCL();
+
+  // Bump collective counter
+  seq_++;
+
+  const auto devices = getDeviceList(tensor);
+  const auto key = getKeyFromDevices(devices);
+  auto& ncclComms = getNCCLComm(key, devices, opType);
+
+  syncStreams(devices, ncclEvents_[key], ncclStreams_[key]);
+
+  // Work itself will create the CUDA events on all GPUs of tensors
+  bool can_profile = tensor.size() == 1;
+  auto work = initWork(
+      devices,
+      rank_,
+      opType,
+      can_profile ? profilingTitle : nullptr,
+      can_profile ? c10::optional<std::vector<at::Tensor>>(tensor)
+                  : c10::nullopt);
+
+  // Store references to outputs to be used by WorkNCCL::result and operator<<.
+  if (opType == OpType::GATHER) {
```

Comment:
Why do we have this conditional here?

Reply:
This is because for Gather and Scatter, we are actually defining `outputs` differently, for `gather`, `outputs` is `tensor_list`, while for `scatter`, `outputs` is defined as `tensor`